Mesh light Merge

danieldresser opened the pull request:
Presenting my merge of the Mesh Light branch for reference.  Nothing significant here, just 3 commits:
* one merge conflict
* one manual merge of change to m_node from instancing commits
* fix for ArnoldMeshLight stomping light set.

I'm pretty confident with you merging this for the next major version.  I've got all the previous lights and light visualisers in IERendering working with this code now, and I'm in the middle of getting mesh lights set up in IERendering.  Test renders of mesh lights in Arnold look fine.

johnhaddon commented:
Thanks Daniel - any chance you could rebase to remove the funny merge commit in the middle?

danieldresser commented:
Did a rebase where I folded all fixes into the original commits

danieldresser commented:
After some internal IE discussion, we don't want to merge this yet.  We want to get out a 0.26 release with the fix you're working on John, and then I guess this will be 0.27.

johnhaddon commented:
Is there anything I need to address with it, or are we just wanting to limit the amount of change in this version?

andrewkaufman commented:
Its mainly just that @danieldresser is on holidays all next week, and this one is breaking to his IELights, so we don't want to release them until he's back on the box.

johnhaddon commented:
I've taken this off on-hold, since @danieldresser is back now, and I'd like to get this stuff rolling.

danieldresser commented:
Rebased to current trunk, and added one more commit you should take a look at John: skipping creating mesh_light with NULL meshes when you filter a mesh light creator over locations that include some empty locations ( or non-leaf locations ).

There is still one issue that I'm not sure how you want to fix.  Currently light shaders are not created using the shader cache.  This means that if you assign the same light shader with the same texture to multiple meshes, you get this error:

00:00:00   290MB WARNING | can't rename node "" to "shader0e9824a95a83ebe176b34051a8afd15e_1_image1": name already in use

During renders this is fairly innocuous, but it does cause corruption if you try and export an .ass file.

The easiest fix I can think of is that IEGafferArnoldRendering::ArnoldShader::ArnoldShader could take an argument that would override the namePrefix, and when outputting lights, the name prefix should include the hash of the location?  This should be adequate to prevent overlap.

It would be better if individual shader components such as image nodes could be shared between lights if they are identical, but I guess this isn't how the shader cache works currently, and maybe sharing pieces of shaders would make interactive updates harder or something?

johnhaddon commented:
Closing in favour of the slightly updated #1787.
